### PR TITLE
Remove unnecessary includes

### PR DESCRIPTION
**Pull request by tautschnig**
Manual removal of includes based on include-what-you-use's output, filtered for includes that should be removed. The goal is to avoid unnecessary build dependencies, reducing the amount of code that needs to be rebuilt during incremental builds.

This is a second iteration of this cleanup; the first round took place in 450845dbc5.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]

**Review by kroening — APPROVED** [review]

**Review by TGWDB — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]
Thanks for this clean up!

**Comment by tautschnig** [comment]
@chris-ryder Could you please take a look at this one?

**Review by chris-ryder — APPROVED** [review]
Looks like a painful, but very useful cleanup!

**Comment by TGWDB** [comment]
@tautschnig  As a member of @diffblue/diffblue-opensource I approved before when it passed checks - is something else needed?

**Comment by tautschnig** [comment]
> @tautschnig As a member of @diffblue/diffblue-opensource I approved before when it passed checks - is something else needed?

I've now added a GitHub action to run include-what-you-use, I believe this triggered the additional review request.

File: src/solvers/smt2_incremental/smt_response_validation.h
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_response_validation.h:8]: `response_or_errort` handles the type it is specialised with by-value like optional does. `response_or_errort` is specialised with `smt_responset` for the return value of `validate_smt_response`. Therefore I would say that it is incorrect to remove this include as it is used.

File: src/solvers/smt2_incremental/smt_to_smt2_string.cpp
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_to_smt2_string.cpp:16]: `std::stringstream` is used in the `smt_to_smt2_string` function (all overloads) and is defined in the `<sstream>`  header. Therefore this include should not be removed, as it is used. I assume this file still compiles without the include due to transitive includes of some type. My understanding is that some of these transitive includes can be in implementations of the STL but not required by the standard.

File: src/solvers/smt2_incremental/type_size_mapping.h
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/type_size_mapping.h:11]: `smt_termt`is defined in `smt_terms.h` and is used in the definition of `type_size_mapt` in this file. Therefore this include should be kept.

File: unit/solvers/smt2_incremental/smt_to_smt2_string.cpp
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/smt_to_smt2_string.cpp:6]: The logics defined in this header are used as part of the "Test smt_set_logic_commandt to string conversion" test in this file. Removing the include still compiles as it depends on a transitive include via `smt_commends.h`.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

File: src/solvers/smt2_incremental/smt_response_validation.h
Comment (tautschnig) [reply src/solvers/smt2_incremental/smt_response_validation.h:8]: I have now added `// IWYU pragma: keep` instead.

File: src/solvers/smt2_incremental/smt_to_smt2_string.cpp
Comment (tautschnig) [reply src/solvers/smt2_incremental/smt_to_smt2_string.cpp:16]: I have now added `// IWYU pragma: keep` instead.

File: src/solvers/smt2_incremental/type_size_mapping.h
Comment (tautschnig) [reply src/solvers/smt2_incremental/type_size_mapping.h:11]: I have now added `// IWYU pragma: keep` instead.

File: unit/solvers/smt2_incremental/smt_to_smt2_string.cpp
Comment (tautschnig) [reply unit/solvers/smt2_incremental/smt_to_smt2_string.cpp:6]: I have now added `// IWYU pragma: keep` instead.

**Review by thomasspriggs — APPROVED** [review]

File: src/goto-symex/symex_main.cpp
Comment (thomasspriggs) [inline src/goto-symex/symex_main.cpp]: I am not convinced this commit is actually using any copy-elision. Are the performance characteristics still the same?

Comment (tautschnig) [reply src/goto-symex/symex_main.cpp]: Actually I should not have included this in here, it was just drive-by cleanup. Commit now removed and instead PR-ed as #7275. Perhaps you can elaborate on that new PR?

**Review by martin-cs — APPROVED** [review]
All areas I am maintainer for look good.

**Pull request merged**
